### PR TITLE
feat(GCS+gRPC): implement RewriteObject()

### LINKED ISSUE #4193: Implement storage::GrpcClient::RewriteObject

### PR DESCRIPTION
**Pull request by coryan**
Fixes #4193

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8100)

### CODE REVIEW

File: google/cloud/storage/internal/grpc_object_request_parser.cc [hunk 1]
```diff
@@ -360,6 +371,101 @@ ListObjectsResponse GrpcObjectRequestParser::FromProto(
...
+    destination.set_cache_control(m.cache_control());
+    destination.set_content_language(m.content_language());
+    destination.set_content_type(m.content_type());
+    destination.set_temporary_hold(m.temporary_hold());
+    for (auto const& kv : m.metadata()) {
+      (*destination.mutable_metadata())[kv.first] = kv.second;
+    }
+    if (m.event_based_hold()) {
+      // The proto is an optional<bool>, avoid setting it to `false`, seems
+      // confusing.
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc_object_request_parser.cc:401]: Ok so false and unset are not the same things:
https://github.com/googleapis/googleapis/blob/ce08228300f5bf6a0c7153176e3adac501eaa5a1/google/storage/v2/storage.proto#L1984-L1986

you could add like a
```
bool ObjectMetadata::use_parent_default_event_based_hold() const;
```

(not saying you should. I'm just doing my due diligence as a reviewer)

**Review by dbolduc — APPROVED** [review]
questions only

File: google/cloud/storage/internal/grpc_object_request_parser.cc [hunk 2]
```diff
@@ -360,6 +371,101 @@ ListObjectsResponse GrpcObjectRequestParser::FromProto(
   return result;
 }
 
+StatusOr<google::storage::v2::RewriteObjectRequest>
+GrpcObjectRequestParser::ToProto(RewriteObjectRequest const& request) {
+  google::storage::v2::RewriteObjectRequest result;
+  SetCommonParameters(result, request);
+  auto status = SetCommonObjectParameters(result, request);
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/grpc_object_request_parser.cc:378]: They are not that "common", `ListObjects()` does not have those parameters.  Maybe they should be called `SetCsekParameters()` but the protos call them "common" :shrug:

File: google/cloud/storage/internal/grpc_object_request_parser.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_object_request_parser.cc:401]: There is an ongoing discussion about how to deal with these optionals.  The protos may need to change to deal with this.  Yes, it is ugly.

File: google/cloud/storage/internal/grpc_object_request_parser.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/storage/internal/grpc_object_request_parser.cc:378]: Thanks, I see `message CommonObjectRequestParams {}` now.

**Pull request merged**